src/main/java/org/kiwiproject/dropwizard/util/startup/StartupLocker.java

```diff
@@ -0,0 +1,136 @@
+
+    @Getter(AccessLevel.PACKAGE)
+    @VisibleForTesting
+    private final CuratorFrameworkHelper curatorFrameworkHelper;
+
+    @Getter(AccessLevel.PACKAGE)
+    @VisibleForTesting
+    private final CuratorLockHelper curatorLockHelper;
+
+    private final SystemExecutioner executioner;
+
+    @Builder
+    private StartupLocker(SystemExecutioner executioner,
+                          ZooKeeperAvailabilityChecker zkAvailabilityChecker,
+                          CuratorFrameworkHelper curatorFrameworkHelper,
+                          CuratorLockHelper curatorLockHelper) {
+        this.executioner = requireNotNull(executioner, "SystemExecutioner is required");
+        this.zkAvailabilityChecker = Optional.ofNullable(zkAvailabilityChecker).orElse(new ZooKeeperAvailabilityChecker());
+        this.curatorFrameworkHelper = Optional.ofNullable(curatorFrameworkHelper).orElse(new CuratorFrameworkHelper());
+        this.curatorLockHelper = Optional.ofNullable(curatorLockHelper).orElse(new CuratorLockHelper());
+    }
+
+    @AllArgsConstructor
+    public static class StartupLockInfo {
+        final CuratorFramework client;
+        final InterProcessLock lock;
+        final String lockPath;
+    }
+
+    /**
+     * Attempts to acquire a lock from ZooKeeper during startup.
+     *
+     * @param lockPath      the path in ZooKeeper to store the lock
+     * @param lockTimeout   the amount of time to wait for the lock to be acquired
+     * @param curatorConfig the Curator configuration
+     * @param environment   the Dropwizard environment
+     * @return information about the acquired lock or null if lock could not be acquired
+     */
+    public StartupLockInfo acquireStartupLock(String lockPath, Duration lockTimeout, PortAssignment assignment, CuratorConfig curatorConfig, Environment environment) {
+        if (assignment == PortAssignment.STATIC) {
+            return null;
+        }
+
+        if (zkAvailabilityChecker.anyZooKeepersAvailable(curatorConfig)) {
+            var curatorFramework = curatorFrameworkHelper.startCuratorFramework(curatorConfig);
+            var lock = curatorLockHelper.createInterProcessMutex(curatorFramework, lockPath);
+
+            if (tryAcquireStartupLock(lock, lockPath, lockTimeout)) {
+                environment.lifecycle().addLifeCycleListener(new StartupWithLockJettyLifeCycleListener(curatorFramework, lock, lockPath, executioner));
+                return new StartupLockInfo(curatorFramework, lock, lockPath);
+            } else {
+                LOG.warn("Lock on path [{}] not obtained. Closing Curator.", lockPath);
+                curatorFrameworkHelper.closeQuietly(curatorFramework);
+            }
+        } else {
+            LOG.warn("No ZooKeepers are available from connect string [{}]", curatorConfig.getZkConnectString());
+        }
+
+        LOG.warn("Startup using dynamic ports will continue without ZooKeeper lock (which may result in port conflicts)");
```

Comment:
This comment isn't really accurate when in this standalone library...since we cannot know what action the calling client code will or will not take. Suggest removing it or changing it to something more generic.

Reply:
With my refactoring from the above comment, this is OBE